K8SPSMDB-1654 Documented support for vector search (#361)
percona/k8spsmdb-docs: docs/vector-search-setup.md

# Configure vector search

!!! admonition "Version added: [1.23.0](RN/Kubernetes-Operator-for-PSMONGODB-RN1.23.0.md)"

!!! warning "Tech preview"

    Vector search is in the tech preview stage. Use it in staging or testing
    environments and share your feedback. 

This guide shows how to enable vector search on a cluster managed by
Percona Operator for MongoDB, insert sample vector data, create a vector search
index, and run a `$vectorSearch` query.

This setup uses the following software versions:

* Percona Operator for MongoDB 1.23.0
* Percona Server for MongoDB 8.3

> * mongot 1.70.1.

Reviewer: what's this version? 
```
$ docker run -it --rm perconalab/percona-server-mongodb-mongot:0.51.0 bash
[mongodb@1690e8276f63 /]$ mongot --version
0.51.0
```

    ```

## Deploy the Operator

Install the Operator by applying the `deploy/bundle.yaml` manifest. This also installs CRDs, Role-based access control (RBAC) and the Operator deployment:
    
```bash
kubectl apply --server-side -f deploy/bundle.yaml
```

As the result you will have the Operator Pod up and running.

## Install Percona Server for MongoDB and enable vector search

1. Edit the `deploy/cr.yaml` Custom Resource. Specify the following keys:

    * Set the database image to Percona Server for MongoDB 8.3 in `spec.image`
    * Set `spec.search.enabled` to `true`
    * Specify the `mongot` image in `spec.search.image`
    * Keep `size: 1` as only one `mongot` pod is allowed per replica set or shard.
    * Configure storage for `mongot` based on your data set. Check [Requirements](vector-search.md#availability-and-requirements) for guidance on estimating storage size.

    Here's the example configuration:

    ```yaml
    apiVersion: psmdb.percona.com/v1
    kind: PerconaServerMongoDB
    metadata:
      name: some-name
      finalizers:
        - percona.com/delete-psmdb-pvc
    spec:
      image: perconalab/percona-server-mongodb:8.3
      search:
        enabled: true

>         image: percona/percona-server-mongodb-mongot:1.70.1

Reviewer: shouldn't we set an existing image here? `perconalab/percona-server-mongodb-mongot:0.51.0`

Author: I will replace it with the variable to update it everywhere shoud the version change.

> # Configure vector search

Reviewer 2: ```suggestion
# Configure Percona Search for MongoDB
```

!!! admonition "Version added: [1.23.0](RN/Kubernetes-Operator-for-PSMONGODB-RN1.23.0.md)"

!!! warning "Tech preview"

>     Vector search is in the tech preview stage. Use it in staging or testing

Reviewer 2: ```suggestion
    Percona Search for MongoDB is in Tech Preview. It's not recommended for production environments yet.
```

    environments and share your feedback. 

This guide shows how to enable vector search on a cluster managed by
Percona Operator for MongoDB, insert sample vector data, create a vector search
index, and run a `$vectorSearch` query.

This setup uses the following software versions:

* Percona Operator for MongoDB 1.23.0
* Percona Server for MongoDB 8.3

> * mongot {{mongot}}.

Reviewer 2: ```suggestion
* Percona Search for MongoDB {{mongot}}.
```

> To learn how vector search works with the Operator, see

Reviewer 2: ```suggestion
To learn how Percona Search for MongoDB works with the Operator, see
```

> [Vector search](vector-search.md).

Reviewer 2: ```suggestion
[Percona Search for MongoDB](vector-search.md).
```

## Before you start

> 1. Make sure you understand [requirements](vector-search.md#availability-and-requirements) and [limitations](vector-search.md#limitations) of using vector search.

Reviewer 2: ```suggestion
1. Make sure you understand [requirements](vector-search.md#availability-and-requirements) and [limitations](vector-search.md#limitations) of using Percona Search for MongoDB.
```

2. Clone the repository with all manifests and source code:
  
    ```bash
    git clone -b v{{release}} https://github.com/percona/percona-server-mongodb-operator
    cd percona-server-mongodb-operator
    ```

3. Create the namespace and export it as environment variable. Replace the `<namespace>` placeholder with your value:

    ```bash
    kubectl create namespace <namespace>
    export NAMESPACE=<namespace>
    ```

## Deploy the Operator

Install the Operator by applying the `deploy/bundle.yaml` manifest. This also installs CRDs, Role-based access control (RBAC) and the Operator deployment:
    
```bash
kubectl apply --server-side -f deploy/bundle.yaml
```

As the result you will have the Operator Pod up and running.

> ## Install Percona Server for MongoDB and enable vector search

Reviewer 2: ```suggestion
## Install Percona Server for MongoDB and Percona Search for MongoDB
```

    Here's the example configuration:

    ```yaml
    apiVersion: psmdb.percona.com/v1
    kind: PerconaServerMongoDB
    metadata:
      name: some-name
      finalizers:
        - percona.com/delete-psmdb-pvc
    spec:
      image: perconalab/percona-server-mongodb:8.3
      search:
        enabled: true
        image: perconalab/percona-search-mongodb:{{mongot}}
        size: 1
        storage:
          persistentVolumeClaim:
            resources:
              requests:
                storage: 10Gi
        resources:
          requests:
            cpu: "2"
            memory: 2Gi
      replsets:
        - name: rs0
          affinity:
            antiAffinityTopologyKey: none
          resources:
            limits:
              cpu: "500m"
              memory: 1G
            requests:
              cpu: "100m"
              memory: 100Mi
          volumeSpec:
            persistentVolumeClaim:
              resources:
                requests:
                  storage: 3Gi
     #The rest of your configuration
    ```

2. Apply the Custom Resource to install Percona Server for MongoDB:

    ```bash
    kubectl apply -f deploy/cr.yaml -n $NAMESPACE
    ```

> ### Verify that search is ready

Reviewer 2: ```suggestion
### Verify that `mongot` is ready
```

      image: perconalab/percona-server-mongodb:8.3
      search:
        enabled: true

>         image: percona/percona-server-mongodb-mongot:{{mongot}}

Reviewer: one thing i don't understand is that we use `perconalab/percona-server-mongodb` for PSMDB but `percona/percona-server-mongodb-mongot` for mongot.

if we're okay with documenting non-existent image, can't we also use `percona/percona-server-mongodb:8.3`?

Author: My bad for not updating the path to `perconalab`. Thank you for pointing it out. Updated. Reg mongot version, it will be 1.70.1

Reviewer: sorry for being a pain in the ass but if, as of today, there's no 1.70.1 image why we're using it in our docs?
